wasm-libruntime/src/workload.rs

```diff
@@ -0,0 +1,166 @@
+//! The Workload module implements retrieving and running WASM workloads.
+
+use crate::Result;
+use std::collections::HashMap;
+use wasmtime::Val;
+
+/// Struct to hold and process the workload
+pub struct Workload {
+    // TODO: think about if we assume that this buffer contains an unencrypted or encrypted workload
+    buf: Vec<u8>,
+    // TODO: more fields for tracking the state?
+}
+
+impl Workload {
+    /// Retrieves the WASM workload from an argument which implements Read
+    pub fn from_reader<R: std::io::Read>(mut r: R) -> Result<Self> {
+        let mut buf = Vec::new();
+        r.read_to_end(&mut buf)?;
+        Ok(Self { buf })
+    }
+
+    /// Get the length of the internal buffer
+    pub fn len(&self) -> usize {
+        self.buf.len()
+    }
+
+    /// Returns `true` if the workload buffer has a length 0
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Runs a WASM workload, invoking the given function with the given arguments, returning the result vector.
+    pub fn run(&self, function_name: &str, args: &[Val]) -> Result<Vec<Val>> {
+        let app_wasm = &self.buf;
+
+        // Wasmtime API and helper variables
+        let engine = wasmtime::Engine::default();
+        let store = wasmtime::Store::new(&engine);
+        let mut module_registry: HashMap<String, wasmtime::Instance> = HashMap::new();
+
+        // Instantiate WASI
+        let wasi = wasmtime_wasi::instantiate_wasi(
+            // preopened_dirs: &[(String, File)],
+            Default::default(),
+            // argv: &[String],
+            Default::default(),
+            // environ: &[(String, String)],
+            Default::default(),
```

Comment:
Just a comment here for longer terms thoughts. When a tenant submits a workload, we probably want them to also supply deployment metadata. This might include things like:
  * argv
  * environ
  * a whitelist of envvars to accept from the host

Reply:
Without having seen this comment, I did start designing a workload specification mechanism. Please see the latest revision of this PR. I'm looking forward to receiving feedback on it!